internal/pkg/gateway/registry/registry.go

```diff
@@ -7,40 +7,218 @@ SPDX-License-Identifier: Apache-2.0
-	// placeholder code for now - just return host peer - allows verification that Evaluate() works!
-	// this will be replaced by a set of endorsers selected from the discovery endorsement plan
-	// for the given channel/chaincode
-	return []gateway.Endorser{reg.localEndorser}
+func (reg *Registry) Endorsers(channel string, chaincode string) ([]endpoint.Endorser, error) {
+	reg.registerChannel(channel)
+
+	var endorsers []endpoint.Endorser
+
+	interest := &dp.ChaincodeInterest{
+		Chaincodes: []*dp.ChaincodeCall{{
+			Name: chaincode,
+		}},
+	}
+
+	descriptor, err := reg.discovery.PeersForEndorsement(common.ChannelID(channel), interest)
+	if err != nil {
+		return nil, err
+	}
+
+	e := descriptor.EndorsersByGroups
+
+	// choose first layout for now - todo implement retry logic
+	if len(descriptor.Layouts) > 0 {
+		layout := descriptor.Layouts[0].QuantitiesByGroup
+		var r []*dp.Peer
+		for group, quantity := range layout {
+			// For now, select n peers from each group at random. Future story, sort by block height
+			endorsers := e[group].Peers
+			rand.Shuffle(len(endorsers), func(i, j int) {
+				endorsers[i], endorsers[j] = endorsers[j], endorsers[i]
+			})
+			peerGroup := endorsers[0:quantity]
+			r = append(r, peerGroup...)
+		}
+
+		for _, peer := range r {
+			msg := &gossip.GossipMessage{}
+			err := proto.Unmarshal(peer.MembershipInfo.Payload, msg)
+			if err != nil {
+				return nil, err
+			}
+			endpoint := msg.GetAliveMsg().Membership.Endpoint
+			if endpoint == reg.selfEndpoint {
+				endorsers = append(endorsers, reg.localEndorser)
+			} else if endorser, ok := reg.peers[endpoint]; ok {
+				endorsers = append(endorsers, endorser)
+			} else {
+				reg.logger.Warnf("Failed to find endorser at %s", endpoint)
+			}
+		}
+	}
+
+	return endorsers, nil
 }
 
 // Orderers returns a set of orderers that can order a transaction for the given channel.
-func (reg *Registry) Orderers(channel string) []orderer.AtomicBroadcast_BroadcastClient {
-	// placeholder code for now - just return empty array - Submit() won't work at the moment.
-	// this will be replaced by the set of orderers for a channel determined from the discovery service
-	return make([]orderer.AtomicBroadcast_BroadcastClient, 0)
+func (reg *Registry) Orderers(channel string) ([]endpoint.Orderer, error) {
+	reg.registerChannel(channel)
+	var orderers []endpoint.Orderer
+
+	// Get the config
+	config, err := reg.discovery.Config(channel)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, eps := range config.GetOrderers() {
+		for _, ep := range eps.Endpoint {
+			url := fmt.Sprintf("%s:%d", ep.Host, ep.Port)
+			if ord, ok := reg.orderers[url]; ok {
+				orderers = append(orderers, ord)
+			}
+		}
+	}
+
+	return orderers, nil
+}
+
+func (reg *Registry) addPeer(mspid string, url string) error {
+	// assumes that the msp registry has already been populated with this mspid
+	if _, ok := reg.peers[url]; !ok && strings.Contains(url, ":") {
```

Comment:
This is an interesting check.  What does a `:` signify? I can guess but it seems the significance should be made explicit.

Reply:
ah ok - it seems the local (host) peer identifies itself with an empty string.  I've changed the code to be more explicit and added a comment.